### ISSUE TITLE
Vertico slows down

### ISSUE THREAD

**Issue by sienic** [doomemacs/doomemacs#6622]
### What did you expect to happen?

Upon using vertico, it should work as fast as it is in a fresh start.

### What actually happened?

After using emacs for some time, upon dispatching completion (e.g. M-x) Vertico slows down and it takes more than 2s to show results in the minibuffer.

This is the profiler:

![profiler-vertico](https://user-images.githubusercontent.com/71565592/182190868-b09fe2d7-78c6-4d6a-ad32-8cdc96a4e41d.png)

### Describe your attempts to resolve the issue

I haven't been able.  I tried deactivating the global modes of those functions with the `cmhh` suffix, to no avail.

### Steps to reproduce

1. Open Emacs
2. Use Vertico + Consult
3. Do M-x
4. After some hours of use note how it gets slower and how the profiler will start showing under `minibuffer-mode` (in the cpu profiler) ever more items.

### System Information

https://pastebin.com/ey7yiTy5

**Comment by minad** [comment]
Interesting. There is clearly something wrong if these `-cmhh` functions take 2% each. Can you please post your full profile as a file?

The `-cmhh` (change major mode hook) functions are supposed to be fast since they do nothing else than checking if a certain globalized minor mode should be enabled in a given buffer.  As an alternative if all these globalized modes turn out to be a problem, one could consider to enable the minor modes buffer locally per major mode instead. Generally, globalized minor modes seem like not such a great idea due to their intrusiveness and potential bad interactions with major modes and other minor modes. In my personal setup I minimized their usage.

**Comment by minad** [comment]
For reference - this is the definition of the `-cmhh` function in the `define-globalized-minor-mode` macro. There is nothing expensive going on.

~~~ elisp
       (defun ,MODE-cmhh ()
	 (add-to-list ',MODE-buffers (current-buffer))
	 (add-hook 'post-command-hook #',MODE-check-buffers))
~~~

**Comment by fabian-thomas** [comment]
I'm experiencing the same.

**Comment by brabalan** [comment]
I’m also seeing the same issue. If it matters, I’m using `layout +bepo` (it had some incidence on `M-x` at a point).

**Comment by brabalan** [comment]
I have the impression the issue is with `vertico--exhibit`. When `M-x` takes too long, if I interrupt it, I get a message that it was this function that was interrupted.

**Comment by minad** [comment]
If `vertico--exhibit` is slow, the actual problem is the underlying candidate computation via `completion-all-completions`. However note that Vertico is interruptible. While the computation takes place you can already enter some input and the computation is restarted. Given more input the computation will finish more quickly.

To demonstrate the issue you can install an alternative completion style, which refuses to compute candidates if the input is too short:

~~~ elisp
(defun +orderless-limited-all-completions (string table pred point)
  ;; Refuse to compute candidates if input is too short
  (and (length> string 2)
       (or (orderless-all-completions string table pred point)
           ;; Fallback to prefix completion to support dynamic completion tables
           (completion-emacs21-all-completions string table pred point))))

(defun +orderless-limited-try-completion (string table pred point)
  ;; Refuse to compute candidates if input is too short
  (and (length< string 2)
       (or (orderless-try-completion string table pred point)
           ;; Fallback to prefix completion to support dynamic completion tables
           (completion-emacs21-try-completion string table pred point))))

(add-to-list 'completion-styles-alist
             '(+orderless-limited
               +orderless-limited-try-completion
               +orderless-limited-all-completions
               "Limited orderless completion style."))

(setq completion-styles '(+orderless-limited)
      completion-category-defaults nil
      completion-category-overrides nil)
~~~

**Comment by brabalan** [comment]
Thank you. It does seem to speed things up, at the cost of usability. For instance, when asking for correction of a spelling mistake, the list is blank until I start typing (but then I don’t know the mistake). It’s the same to switch buffers: although the list is not long, it is initially empty, and switching to a recently used buffer is no longer possible unless I remember its name.

**Comment by minad** [comment]
@brabalan Of course, it prevents the candidates from being computed if there is no input. Anyway if your M-x is slow, you may just have too many commands to handle efficiently. On my Emacs 28 I have about 6k commands and 25k functions (C-h f) and Vertico is always fast for such numbers of candidates.

**Comment by brabalan** [comment]
What I find surprising is that it used to be fast, then there was a bug that would freeze everything (fixed at https://github.com/doomemacs/doomemacs/commit/2c14eff7f145040c4df923c9c6546b38fcfdc7ee), and since then it’s slowing down.

I have 11k commands and 47k functions. `M-x` is slow (about 4 seconds), while `SPC h f` is quite fast.

**Comment by minad** [comment]
@brabalan Okay, I see. 11k commands is well within the bounds of what Vertico can handle. Then the issue must be caused by some Doomism, as in the commit https://github.com/doomemacs/doomemacs/commit/2c14eff7f145040c4df923c9c6546b38fcfdc7ee.

**Comment by minad** [comment]
Did you try disabling marginalia-mode? But I wonder, if the docstring loading is the reason for the slowdown then this should also be reflected in the profiles.

**Comment by brabalan** [comment]
Disabling marginalia-mode make the completion instantaneous.  Thank you for the suggestion!

**Comment by dmusican** [comment]
I just stumbled across this bug report; I'm experiencing the same problem.

**Comment by mmajewsk** [comment]
I'm having the same issues when using `SPC ,` for switching buffer. I'm using python, anaconda, and lsp. I'm suspecting that emacs is trying to run some helpers on the buffers, and that this causes the slowdown. The marginalia disabling didn't help.

**Comment by RequiemSouls** [comment]
<img width="604" alt="image" src="https://github.com/doomemacs/doomemacs/assets/3190094/ce46fd14-37b3-4eac-a2b7-f26309c9574e">

the same issue when I just move the cursor use hjkl. I use (cc +lsp) lsp client is eglot. disable eglot can't make the speed change.

**Comment by vale981** [comment]
If you're using emacs pgtk on wayland, try running the "normal" emacs gtk build under xwayland. This improved perfomance significantly in my case.

**Comment by JulienMalka** [comment]
I am indeed using pgtk under wayland

**Comment by oncomouse** [comment]
Just chiming in that I am having similar issues as @RequiemSouls is illustrating with his profiler report. This is happening both on Linux in x11 and on macOS. Disabling `marginalia-mode` does seem to make this go away.

**Comment by Tonelllo** [comment]
I was having a similar problem. When starting emacs it would slowly increase the cpu usage untill reaching 100%. After disabling `marginalia-mode` the problem seems solved. My profiler report is identical to the ones posted above. Ubuntu 23.10 under X11.

**Comment by jskulski** [comment]
A few more data points: 

- disabling marginalia-mode did not solve the issue for me (either in-editor or adding to custom.el)
- i turned off veritco and turned on company mode with the same cpu 100% behavior 
- @minad's elisp code lowers cpu usage w/ helm and vertico (ty!)

cpu spike started happening after a rebuild of the same emacs version, in both daemon and editor processes. 

nixos 24.05 + 23.11 
emacs 28.2 + 29.1

Added a profile over here:
https://gist.github.com/jskulski/9ced615f03a1c0933ce7885cc7060cce

**Comment by hlissner** [comment]
> I was having a similar problem. When starting emacs it would slowly increase the cpu usage untill reaching 100%

Are you sure these CPU issues are related? This could also be explained by native-comp compiling elisp in the background, which it will do after new packages have been installed, old packages have been updated, or you've up/downgraded Emacs itself (rendering old native-bytecode stale).

----

Another possible explanation for all this slowness could be fonts. If there are glyphs displayed anywhere in Emacs, and you don't have its fallback font (`Symbola`) installed (and your active font doesn't support it), Emacs can choke or crash trying to render it. Which could be made worse by Vertico displaying many of these glyphs at once.

In any case, I haven't been able to reproduce this yet (beyond what was fixed in https://github.com/doomemacs/doomemacs/commit/2c14eff7f145040c4df923c9c6546b38fcfdc7ee).

**Comment by Tonelllo** [comment]
> > I was having a similar problem. When starting emacs it would slowly increase the cpu usage untill reaching 100%
> 
> Are you sure these CPU issues are related? This could also be explained by native-comp compiling elisp in the background, which it will do after new packages have been installed, old packages have been updated, or you've up/downgraded Emacs itself (rendering old native-bytecode stale).
> 
> Another possible explanation for all this slowness could be fonts. If there are glyphs Emacs can't render with your active font or its fallback font isn't installed (`Symbola`), Emacs can choke or crash trying to render them. Which could be exacerbated by Vertico displaying many of these glyphs at once.
> 
> In any case, I haven't been able to reproduce this yet (beyond what was fixed in [2c14eff](https://github.com/doomemacs/doomemacs/commit/2c14eff7f145040c4df923c9c6546b38fcfdc7ee)).

You are right, the CPU usage was in fact due to Emacs native-compiling the packages is my case. Sorry for the mistake.
For now I have not experienced any slow down. In case anything changes I will provide updates.

**Comment by hlissner** [comment]
I believe this was fixed in https://github.com/doomemacs/doomemacs/commit/4425bd898f4cd53d777190359a85a3d80f8d01eb (and later, https://github.com/doomemacs/doomemacs/commit/73693a8a2d675bc121f09b476df70e89eaf902bd). `doom-project-root` was called on every item Marginalia processed (which calls `projectile-project-root`, and that gets *very* expensive when `default-directory` is not in a valid project — causing dozens of I/O calls *per* item in the minibuffer).

In any case, I'll assume this is resolved, but feel free to follow up here otherwise.
